### DISCUSSION TITLE
Loop inside list<T> initialization.

### DISCUSSION THREAD

**Discussion by Sharaf-Mansour** [dotnet/csharplang#3993] [Language Ideas] [answered]
I made a list of records and It looks like this.
![image](https://user-images.githubusercontent.com/55330747/95677135-bf3e6e00-0bc3-11eb-8a25-aa3f26a756a9.png)
It goes on for 250 times.
But I wanted to use a loop to make things easier but to do it the only way would be like this.
![image](https://user-images.githubusercontent.com/55330747/95677220-55729400-0bc4-11eb-9157-a1d7def46342.png)
So I thought of having some sort of a way to improve this by doing something like this.
![image](https://user-images.githubusercontent.com/55330747/95677309-e77a9c80-0bc4-11eb-9840-cf448bdd8e25.png)
Just an idea, might be helpful.

**Comment by YairHalberstadt** [comment]
I'm not sure what about your idea is specific to records? Why not any type?

Languages like scala allow expressions to be iterators, allowing something like:

```scala
(for (i <- 1 to 250) yield new Score(i)).ToList();
```
which looks pretty close to what you wrote.

In c# however a `for`/`foreach` loop isn't an expression, so this approach wouldn't work as well.

However you can do this:

```csharp
Enumerable.Range(1, 250).Select(i => new Score(i)).ToList();
```
Which is pretty idiomatic C#.
Would you consider that sufficient?

**Comment by YairHalberstadt** [comment]
BTW instead of copying over screenshots of your code, try copying it over, and placing it between 3 backticks as follows:

\`\`\`csharp
csharp code goes here and is syntax highlighted
\`\`\`

This makes it easier for people to copy and paste your code, and also helps people who use [screen readers](https://en.wikipedia.org/wiki/Screen_reader) :-)

Thanks, and welcome to csharplang !

**Reply by Sharaf-Mansour** [reply]
Thanks for your feedback ❤️ I am still new 😄

**Comment by Sharaf-Mansour** [comment]
@YairHalberstadt I agree it should be to all types.
after testing your code it looked like this.
```csharp
   public static List<Score> scores98_361 =>  Enumerable.Range(1, 250).Select(i => new Score(i)).ToList();
```
it is not the best looking but what is the performance cost from writing 250 lines of code and writing the code above ?
Another issue will be complicated initialization Like this
```csharp
 public static List<Exam> Exams98_361 { get; set; } = new () {
            new  (1, new () {
                new  (1, "Q1/98-361-Q1-A1.jpg", true),
                new  (2, "Q1/98-361-Q1-A2.jpg", false),
                new  (3, "Q1/98-361-Q1-A3.jpg", false),
                new  (4, "Q1/98-361-Q1-A4.jpg", false),
            }),
            new  (2, new () {
                new  (1, "Q2/98-361-Q2-A1.jpg", true),
                new  (2, "Q2/98-361-Q2-A2.jpg", false),
                new  (3, "Q2/98-361-Q2-A3.jpg", true),
                new  (4, "Q2/98-361-Q2-A4.jpg", false),
            }),
            new  (3, new () {
                new  (1, "Q3/98-361-Q3-A1.jpg", false),
                new  (2, "Q3/98-361-Q3-A2.jpg", true),
                new  (3, "Q3/98-361-Q3-A3.jpg", false),
                new  (4, "Q3/98-361-Q3-A4.jpg", false),
                new  (5, "Q3/98-361-Q3-A5.jpg", false),
                new  (6, "Q3/98-361-Q3-A6.jpg", false),
              }),
        };
```
And yes I did write that 250 times in my code

**Reply by huoyaoyuan** [reply]
> but what is the performance cost

LINQ does have performance cost. But generally, in application code, you don't need to worry about it, because the cost is not very much. C# wasn't designed for zero overhead abstraction in previous version. **As an application developer, don't worry about every minor performance impact.**

**Reply by Sharaf-Mansour** [reply]
I am using Blazor WASM So that is why I asked about performance impact, Thanks for your feedback!

**Reply by huoyaoyuan** [reply]
Do an actual benchmark is a good practice. Sometimes benchmark shows surprising result to me.
For WASM, the platform is very new for .NET. Waiting for more internal optimization may give you more improvement than just changing the compiling result.

**Reply by Sharaf-Mansour** [reply]
I have another issue with .net5 RC1 and GitHub pages I can not get it to work 😫 , And I am trying to get my project on azure for benchmark. I have never tried any of this before

**Comment by spydacarnage** [comment]
Anything you could do in your for loop could be done with Enumerable.Range.

**Reply by Sharaf-Mansour** [reply]
I know it could be done I said that it will be complicated considering I have to loop over those 2:

```csharp
public record Exam(int ID, List<Answer> Answers, string Value = "");
public record Answer(int ID, string Value, bool IsCorrect, bool IsSelected = false);
```

**Reply by YairHalberstadt** [reply]
There's a straightforward conversion from anything you can do in a for loop to anything you can do in a linq query. It can be tricky to learn, but will but quite simple once you're used to it. Definitely worth practicing :-)

**Reply by Sharaf-Mansour** [reply]
I will start right away.

**Comment by spydacarnage** [comment]
I think I'm missing why it's more complicated, but if you think it is,
that's a valid opinion.

**Reply by Sharaf-Mansour** [reply]
Because I will have to do a loop inside another loop.
I am just new here 😆 
Edit: Also working with optional parameters

**Comment by HaloFour** [comment]
Or you could write it as multiple statements.  There's nothing wrong with initializing a `List<T>` and then using a loop to populate it.

**Reply by Sharaf-Mansour** [reply]
True, I just think it will be easier to have for / foreach loop expression that work at initializing.
It is just an idea.

**Comment by spydacarnage** [comment]
Which you can still do using LINQ. And depending what you are looping
around, it could be more efficient.

**Reply by Sharaf-Mansour** [reply]
Thanks for your feedback I am testing @Flutterish approach. Which I think is what you are talking about.

**Comment by Flutterish** [comment]
if you prefer a "list comprehention"-ish approach
```cs
var myList = (from i in Enumerable.Range(1, 250) select new MyRecord( i )).ToList();
// or (from i in 1..251 select new MyRecord( i )).ToList(); if range to enumerable is implemeted
```

**Reply by Sharaf-Mansour** [reply]
I will test that right now, Thanks for your feedback!

**Comment by Sharaf-Mansour** [comment, marked as answer]
After testing I think @Flutterish  approach @YairHalberstadt  are the same 
```csharp
 public static List<Score> scores98_361 => Enumerable.Range(1, 250).Select(i => new Score(i)).ToList();
 public static List<Score> scores98_362 => (from i in Enumerable.Range(1, 250) select new Score(i)).ToList();
```
I think having something like this in C# will be neat 😃 
```csharp
 public static List<Score> scores98_361 =  for(var i =1;i<251;i++) {  new Score(i)    };
// or

 public static List<Score> scores98_361 =  for(var i =1;i<251;i++) => new Score(i);

//And for each 

 public static List<int> Numbers =  for(var i =1;i<251;i++) => i;

 public static List<Score> scores98_361 = foreach (var item in Numbers) => new Score(item );

```

**Reply by HaloFour** [reply]
Even if that was to be considered the end result would not be a `List<T>` as that would require evaluating the entire loop eagerly which is pretty much against the design of iterators in general.  LINQ provides all of this functionality and a lot more, I don't see why there'd be a reason to bring it into the language proper, especially without considering all of the additional syntax that would be required to make it capable of filtering, mapping and collecting into one of any number of results or collection types.

**Reply by YairHalberstadt** [reply]
This is actually a really good example to discuss how language design works :-)

Every single thing a developer writes, would of course be quicker if there was a language feature which did exactly that for you. But imagine trying to learn a language like that!!! You would have to learn 100,000 tiny features! And implementing and designing it would be a nightmare - you would have to work out how every single feature interacts with every other one, and make sure none of them share the same syntax or anything like that.

Instead language designers try to work out the minimal set of features, that give you the most power. And it turns out just 3 really powerful features, are enough to provide exactly what you want: lambda expressions, generics, and extensions. And not only that, but using these same 3 features, LINQ is able to express the vast majority of transformations to data that developers use.

Once that's done, there's very little incentive to introduce new syntax for the little special cases - especially as you can effectively add your own methods to cover exactly your cases if you want:

```csharp
public static List<T> For<T>(int start, int end, Func<int, T> func) => Enumerable.Range(start, end - start).Select(i => func(i)).ToList();
public static List<TResult> Foreach<TSource, TResult>(IEnumerable<TSource> source, Func<TSource, TResult> func) => source.Select(i => func(i)).ToList();

public static List<Score> scores98_361 = For(1, 251, i => new Score(i));
public static List<Score> scores98_361 = Foreach(Numbers, item => new Score(item));
```

**Reply by Sharaf-Mansour** [reply]
I am so new to this, This changes everything for me, This is amazing! 
Thank you for your huge help. I think that looks even better than what I expected but what about the performance cost?

**Reply by YairHalberstadt** [reply]
You can avoid some of the performance costs by using a for loop inside the methods, and adding to a list which you initialize to the correct size. If you need more, there's other techniques you can use, but it's best to wait till they're actually needed!

**Reply by Sharaf-Mansour** [reply]
There is a lot to learn. Thanks for your efforts and help 👍 👍 💯

**Comment by spydacarnage** [comment]
Unfortunately, as we've provided a perfectly valid, currently supported, way of achieving this fairly niche-case, I don't think this is going to get much traction as a new language feature.

**Reply by Sharaf-Mansour** [reply]
lets wait for MAUI, I am sure .Net 6 will add a lot of new language features also MVU Pattern.
